### PR TITLE
perf(autolinking)!: prioritise files with `*Package.java|*Package.kt` in `getPackageClassName`

### PR DESCRIPTION
**Pull request by szymonrybczak**
Summary:
---------

This Pull Request changes schema that looks for files to check to find class name, after my tests in various application the match results were included almost in every library in files with `Package` at the end. But our previous logic was going through every file in library with `.kt` and `.java` extension until it found file with `implements ReactPackage`, some libraries can have 100+ files and we were reading them until we matched relevant Regex schema 😅

I also added a fallback which checks files with `*.kt` and `*.java` extensions if it couldn't find `implements ReactPackage` in file with `*Package` at the end, which ensures us that we don't introduce any breaking change. **The output of `config` command stays the same.**

Performance improvement
---------

I did a test in [Expensify](https://github.com/Expensify/App) app, which is an open-source app with 50 libraries containing native code. Before my changes, we were reading 490 files to find the class name. Right now, we only check 79 (still there's a room for improvements).

| | Before | After |
|--------|--------|--------|
| Files read | 490 | 79 |
| Time | ~7.85s | ~**1.15s** | 

Test Plan:
----------

1. Clone the repository and do all the required steps from the [Contributing guide](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md)
2. Run this command:

```sh
node /path/to/react-native-cli/packages/cli/build/bin.js config
```
The `dependencies` field should be the same as before. 

Checklist
----------

- [x] Documentation is up to date to reflect these changes.
- [x] Follows commit message convention described in [CONTRIBUTING.md](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md#commit-message-convention)

### CODE REVIEW

File: packages/cli-platform-android/src/config/findPackageClassName.ts
Comment (tido64) [inline packages/cli-platform-android/src/config/findPackageClassName.ts:46]: I wonder if streaming is faster than reading the whole file every time: https://tangledeveloper.medium.com/how-to-read-file-line-by-line-efficiently-in-node-js-8125fe5bd185

Comment (thymikee) [reply packages/cli-platform-android/src/config/findPackageClassName.ts:46]: worth to measure definitely, these declarations are usually somewhere at the top of the file so it hints some potential

Comment (tido64) [reply packages/cli-platform-android/src/config/findPackageClassName.ts:46]: Replacing `glob.sync` above with [`glob.stream`](https://github.com/mrmlnc/fast-glob?tab=readme-ov-file#stream) might also be worth looking into, but making everything async might turn this into a very big PR 😛

File: packages/cli-platform-android/src/config/findPackageClassName.ts
Comment (satya164) [inline packages/cli-platform-android/src/config/findPackageClassName.ts:36]: maybe there's scope to exclude previously matched files from this list not to read same files twice

Comment (szymonrybczak) [reply packages/cli-platform-android/src/config/findPackageClassName.ts:36]: nice! good catch

**Review by cortinico — APPROVED** [review]
Great improvement 👍 All good on my end but wait for another CLI reviews before merging this (also because I believe this is not E2E tested)

File: packages/cli-platform-android/src/config/findPackageClassName.ts
Comment (tido64) [inline packages/cli-platform-android/src/config/findPackageClassName.ts:36]: A bit late to the game but I wonder if globbing for `**/*.{java,kt}` once and filtering in-memory instead of hitting the disk twice is faster for bigger projects.

Comment (thymikee) [reply packages/cli-platform-android/src/config/findPackageClassName.ts:36]: @szymonrybczak keen to measure it and sending a followup if that helps with perf? :)

Comment (szymonrybczak) [reply packages/cli-platform-android/src/config/findPackageClassName.ts:36]: Hm, I gave it a shot but the improvement won't be big because after my checks the amount of libraries that needs to fallback are very small.

Now: 
```
❯ hyperfine "npx react-native config --platform android"
Benchmark 1: npx react-native config --platform android
  Time (mean ± σ):      2.070 s ±  0.264 s    [User: 1.049 s, System: 0.607 s]
  Range (min … max):    1.916 s …  2.814 s    10 runs
```
With executing glob once and filtering in-memory:
```
❯ hyperfine "npx react-native config --platform android"
Benchmark 1: npx react-native config --platform android
  Time (mean ± σ):      2.096 s ±  0.190 s    [User: 1.078 s, System: 0.620 s]
  Range (min … max):    1.931 s …  2.572 s    10 runs
```

**Pull request merged**
